File: newspeedv2/engine.py
import numpy as np
import threading
from pydub import AudioSegment

from mod_transport import TransportDynamics
from mod_magnetic import MagneticPath
from mod_electronics import ElectronicComponents

class TapeEngine:
    def __init__(self, seed=None):
        self.audio_data    = None
        self.total_samples = 0
        self.play_head     = 0.0
        self.current_time  = 0.0
        self.is_playing    = False
        self.is_reversed   = False
        self.lock          = threading.Lock()

        self.transport   = TransportDynamics(seed=seed)
        self.magnetic    = MagneticPath()
        self.electronics = ElectronicComponents()

        self.params        = {}
        self._scrape_phase = 0.0
        self._scrape_state = np.zeros(4)   # [y_prev, x_prev] per channel for HP IIR
        self._os_context   = None          # last N samples before upsample (FIR overlap context)
        self._fade_in      = 0             # samples remaining in post-preset-change crossfade
        self._last_out     = np.zeros(2, dtype=np.float32)  # last output sample for DC crossfade

    def make_worker_engine(self, start_sample, block_size, oversample,
                           shared_seed, n_warmup_blocks=16):
        """
        Create a worker TapeEngine pre-warmed to render from start_sample.

        The engine is initialised with the same oscillator seed as the main
        engine so wow/flutter LFO phases are continuous across slice boundaries.
        It then runs n_warmup_blocks of actual audio (from before start_sample)
        through the full DSP chain and discards the output.  This settles all
        IIR filter states (head bump, azimuth lowpass, pink noise, scrape flutter,
        azimuth delay buffer) to the values they would have had had the engine
        been running continuously from the beginning of the file.

        Worker 0 starts at sample 0 so its warm-up window is empty — it just
        uses the cold-started engine directly (same as single-thread).
        """
        eng = TapeEngine(seed=shared_seed)
        with self.lock:
            eng.audio_data    = self.audio_data
            eng.total_samples = self.total_samples
            eng.params        = dict(self.params)
            eng.is_reversed   = self.is_reversed

        # Warm-up: start engine BEFORE the slice boundary so IIR filters settle.
        warmup_samples = n_warmup_blocks * block_size
        warmup_start   = max(0, start_sample - warmup_samples)
        eng.play_head    = float(warmup_start)
        eng.current_time = float(warmup_start) / 44100.0

        # Motor is fully running mid-tape — no spin-up ramp.
        eng.transport.motor_engage        = 1.0
        eng.transport.current_motor_speed = 1.0

        # Run warm-up blocks and discard their output.
        while eng.play_head < start_sample:
            blk = eng.dsp_process(block_size, oversample=oversample)
            if blk is None:
                break

        # After warm-up, play_head may have overshot start_sample slightly
        # (by up to block_size samples).  Reset it exactly so the worker
        # begins its output at the correct position.
        eng.play_head    = float(start_sample)
        eng.current_time = float(start_sample) / 44100.0
        # Keep _last_out from warm-up so the DC crossfade blends correctly.

        return eng

    def load_file(self, path):
        raw = AudioSegment.from_file(path)
        # pydub only supports mono<->stereo in set_channels, so collapse >2 ch to mono first
        if raw.channels > 2:
            raw = raw.set_channels(1)
        audio = raw.set_frame_rate(44100).set_channels(2).set_sample_width(2)
        audio = audio.apply_gain(-audio.max_dBFS - 1.0)
        samples = np.array(audio.get_array_of_samples()).reshape((-1, 2))
        with self.lock:
            self.audio_data    = samples.astype(np.float32) / 32768.0
            self.total_samples = len(self.audio_data)
            self.is_reversed   = False
            self.reset_state()

    def set_reverse(self, want_reverse, force_reset=False):
        """Flip audio_data in place. Only resets DSP state when direction actually changes."""
        with self.lock:
            if want_reverse != self.is_reversed:
                self.audio_data  = self.audio_data[::-1].copy()
                self.is_reversed = want_reverse
                self.reset_state()
            elif force_reset:
                self.reset_state()

    def reset_state(self):
        self.play_head    = 0.0
        self.current_time = 0.0
        self._os_context  = None
        self.transport.reset()
        self.electronics.reset()
        self.magnetic.reset()

    def dsp_process(self, frames, oversample=1):
        """
        oversample=1  : realtime quality (1024 frames → 1024 output samples)
        oversample=2  : 2× oversampling of nonlinear stages, then decimate
        oversample=4  : 4× oversampling — catches high-order intermod in saturation
        oversample=8  : 8× — diminishing returns but theoretically cleanest
        """
        with self.lock:
            if self.audio_data is None:
                return None
            if self.play_head >= self.total_samples - frames:
                return None

            p            = self.params
            p            = dict(p, is_reversed=self.is_reversed,
                                   motor_engage=1.0)   # fully engaged during playback
            speed_factor = p.get('ips_base', 15.0) / 15.0

            _result = self.transport.process_speed(
                frames, self.current_time, self.play_head, self.total_samples, p
            )
            # Support both old (2-value) and new (3-value) transport return
            if len(_result) == 3:
                final_speeds, sticky_drag, dropout_mask = _result
            else:
                final_speeds, sticky_drag = _result
                dropout_mask = np.ones(frames)

            read_indices = self.play_head + np.cumsum(final_speeds)
            if read_indices[-1] >= self.total_samples - 1:
                return None

            i0    = np.floor(read_indices).astype(np.int32)
            frac  = read_indices - i0
            out_v = self.audio_data[i0] + (self.audio_data[i0+1] - self.audio_data[i0]) * frac[:, None]

            if oversample > 1:
                # Oversampled saturation: upsample → saturate → decimate.
                # resample_poly uses a linear-phase FIR filter. Without context
                # samples from the previous block, the filter cold-starts at
                # block boundaries and produces a transient of up to 0.22 amplitude
                # — audible as a regular click every block_size/44100 seconds.
                # Fix: prepend _os_context (last PAD samples) before upsampling,
                # strip after downsampling so output length stays == frames.
                from scipy.signal import resample_poly
                OS_PAD = 64   # 1.45ms context — enough for all supported OS ratios

                oxide_name = p.get('oxide_type', 'Fe2O3')
                oxide      = self.magnetic.OXIDE_PRESETS.get(
                                 oxide_name, self.magnetic.OXIDE_PRESETS['Fe2O3'])
                drive      = p.get('drive', 1.2)
                bias       = p.get('bias', 1.0) * oxide['bias_trim']
                hc_ratio   = oxide['Hc'] / self.magnetic._Hc_REF
                Ms         = oxide['Ms']
                softness   = float(np.clip(bias, 0.25, 4.0))
                knee_scale = 1.0 / float(np.clip(hc_ratio ** 0.5, 0.5, 4.0))
                ceiling    = Ms / max(drive * 0.5 + 0.5, 0.1)

                # Pad with previous-block context
                has_context = self._os_context is not None
                if has_context:
                    padded = np.vstack([self._os_context, out_v])
                else:
                    padded = out_v   # first block: cold start is inaudible (silence→signal)
                self._os_context = out_v[-OS_PAD:].copy()

                up      = resample_poly(padded, oversample, 1, axis=0)
                h_up    = up * drive
                sat_up  = np.tanh(h_up * knee_scale / softness) * softness / knee_scale * ceiling
                sat_dec = resample_poly(sat_up, 1, oversample, axis=0)

                # Strip the padded prefix so output length == frames
                pad_out = OS_PAD if has_context else 0
                sat_dec = sat_dec[pad_out : pad_out + frames]

                # Strip the padded prefix (OS_PAD input samples → OS_PAD output samples)
                pad_out = 0 if padded is out_v else OS_PAD
                sat_dec = sat_dec[pad_out : pad_out + frames]

                p_os = dict(p, _presaturated=True)
                proc = self.magnetic.process(sat_dec, self.audio_data, read_indices, p_os)
                proc = self.electronics.process(proc, frames, self.current_time,
                                                speed_factor, sticky_drag, p)
            else:
                proc = self.magnetic.process(out_v, self.audio_data, read_indices, p)
                proc = self.electronics.process(proc, frames, self.current_time,
                                                speed_factor, sticky_drag, p)

            # ── Scrape flutter — audio-domain stick-slip modulation ──────────
            # Real scrape flutter: the tape micro-sticks and releases against the
            # stationary erase/guide heads, creating rapid irregular amplitude
            # and phase bursts correlated with high-frequency content.
            # Modelled as: band-limited noise envelope × signal, at a fundamental
            # frequency set by IPS (faster tape → higher scrape frequency).
            #
            # IMPORTANT: after oversampling, proc may be shorter than `frames`
            # (the last block of a pre-roll silence buffer or a speed->1 read
            # can produce fewer output samples).  All array sizes below are
            # derived from len(proc) — never from the `frames` parameter —
            # so we never index out of bounds.
            n_out = len(proc)
            scrape_amt = p.get('scrape_flutter', 0.0)
            if scrape_amt > 0.001:
                SR_f  = 44100.0
                ips   = p.get('ips_base', 15.0)
                scrape_hz  = float(np.clip(2800.0 * (ips / 15.0), 600, 14000))
                phase_inc  = scrape_hz / SR_f
                phases     = self._scrape_phase + np.arange(n_out) * phase_inc
                self._scrape_phase = float(phases[-1] % 1.0)

                env = (np.sin(2 * np.pi * phases)
                     + np.sin(2 * np.pi * phases * 1.031 + 0.7) * 0.6
                     + np.sin(2 * np.pi * phases * 0.973 + 1.3) * 0.4
                     + np.random.normal(0, 0.3, n_out))
                env = np.abs(env) / (np.max(np.abs(env)) + 1e-9)

                alpha  = float(np.clip(0.85 + scrape_amt * 0.1, 0.85, 0.97))
                hf     = np.zeros(n_out)
                y_prev = self._scrape_state[0]
                x_prev = self._scrape_state[1]
                for i in range(n_out):
                    x_cur  = proc[i, 0]
                    y_cur  = alpha * (y_prev + x_cur - x_prev)
                    hf[i]  = y_cur
                    x_prev = x_cur
                    y_prev = y_cur
                self._scrape_state[0] = y_prev
                self._scrape_state[1] = x_prev

                depth = float(np.clip(scrape_amt * 0.4, 0.0, 0.45))
                mod_l = hf * env * depth
                proc[:, 0] += mod_l
                proc[:, 1] += mod_l * 0.85

            # Apply dropout mask — trim to n_out in case transport generated
            # a full-length mask but proc is shorter after OS decimation.
            proc *= dropout_mask[:n_out, None]

            # Fighting speed: when boost and drag conflict, add tape tension AM.
            conflict = getattr(self.transport, 'last_conflict', 0.0)
            if conflict > 0.05:
                t_arr = self.current_time + np.arange(n_out) / 44100.0
                am_freq  = 2.0 + conflict * 3.0
                am_depth = np.clip(conflict * 0.7, 0.0, 0.8)
                am_env   = 1.0 - am_depth * (0.5 + 0.5 * np.sin(2 * np.pi * am_freq * t_arr))
                proc    *= am_env[:, None]

            self.play_head     = float(read_indices[-1])
            self.current_time += frames / 44100.0

            # ── Analog-style output limiter ───────────────────────────────────
            KNEE = 0.97
            proc = np.tanh(proc / KNEE) * KNEE

            # ── Post-preset-change DC crossfade ───────────────────────────────
            # When params change (preset switch or slider move), IIR filter states
            # are tuned to the old preset. The new block's first sample can jump
            # discontinuously. We crossfade from _last_out (the true last output
            # sample) to the new signal over _FADE_SAMPLES samples:
            #   output[n] = new[n]*t + last_out*(1-t)  where t: 0→1
            # At n=0: output = last_out exactly → zero jump across the boundary.
            _FADE_SAMPLES = 512  # ~12ms — covers any IIR settling transient
            if self._fade_in > 0:
                fade_len  = min(self._fade_in, n_out)
                pos_start = _FADE_SAMPLES - self._fade_in
                t_ramp    = np.linspace(pos_start / _FADE_SAMPLES,
                                        (pos_start + fade_len) / _FADE_SAMPLES,
                                        fade_len, dtype=np.float32)
                proc[:fade_len] = (proc[:fade_len] * t_ramp[:, None]
                                   + self._last_out[None, :] * (1.0 - t_ramp[:, None]))
                self._fade_in -= fade_len

            self._last_out = proc[-1].copy()
            return proc

File: nagra-v-cpp/nagra/src/engine.cpp
#include <cstdio>
#include "engine.hpp"
#include <sndfile.h>
#include <cmath>
#include <algorithm>
#include <stdexcept>
#include <cassert>
#include <vector>
#include <random>

TapeEngine::TapeEngine(uint64_t seed)
    : transport(seed), magnetic(), electronics()
{
 }

bool TapeEngine::load_file(const std::string& path) {
    SF_INFO info{};
    SNDFILE* sf = sf_open(path.c_str(), SFM_READ, &info);
    if (!sf) return false;

    // Read raw samples
    std::vector<float> raw(info.frames * info.channels);
    sf_count_t got = sf_readf_float(sf, raw.data(), info.frames);
    sf_close(sf);
    if (got <= 0) return false;

    // Resample to 44100 if needed using SRC_LINEAR via libsndfile virtual IO
    // For simplicity: load as-is and warn if rate != 44100
    // Full SRC resampling would require libsamplerate; we assume 44100 input
    // (the Python code also assumed 44100 after pydub conversion)

    // Convert to stereo Frame[] normalised to ±1.0
    int ch = info.channels;
    int frames_in = (int)got;

    std::lock_guard<std::mutex> g(lock);
    audio_data.resize(frames_in);

    float peak = 1e-9f;
    for (int i = 0; i < frames_in; ++i) {
        float l = raw[i * ch];
        float r = (ch > 1) ? raw[i * ch + 1] : l;
        audio_data[i] = {l, r};
        peak = std::max(peak, std::max(std::abs(l), std::abs(r)));
    }

    // Normalise: apply gain so peak = 0.707 (-3 dBFS, matching pydub -1 dBFS headroom)
    float gain = 0.707f / peak;
    for (auto& f : audio_data) f *= gain;

    total_samples = frames_in;
    is_reversed   = false;
    params.is_reversed = false;
    reset_state();
    reset_position();
    return true;
}

void TapeEngine::set_reverse(bool want_reverse, bool force_reset) {
    std::lock_guard<std::mutex> g(lock);
    // audio_data is ALWAYS in forward order — never physically reversed.
    // is_reversed just tells dsp_process to read indices backward.
    // play_head always means "position in the original forward file".
    is_reversed = want_reverse;
    params.is_reversed = want_reverse;
    if (force_reset) reset_state();
}

void TapeEngine::reset_state() {
    // NOTE: play_head and current_time are intentionally NOT reset here.
    // reset_state() only clears DSP filter state so IIR filters, scrape
    // oscillators and motor state start clean — position is preserved.
    // Call reset_position() separately if you need to seek to zero.
    _os_context.clear();
    _os_context_valid = false;
    _fade_in          = 0;
    _last_out         = {};
    _scrape_phase     = 0.0f;
    _scrape_y_prev    = 0.0f;
    _scrape_x_prev    = 0.0f;
    transport.reset();
    electronics.reset();
    magnetic.reset();
}

void TapeEngine::reset_position() {
    play_head    = 0.0;
    current_time = 0.0f;
}

// ── Oversampled saturation helper ─────────────────────────────────────────────
void TapeEngine::_saturate_oversampled(Frame* buf, int frames, int oversample,
                                        const EngineParams& p)
{
    // Simple polyphase upsample (linear interpolation), saturate, downsample.
    // For production quality use a proper windowed-sinc FIR; this is fast enough
    // for 2–4× and meets the project's latency goals.
    int up_n = frames * oversample;
    std::vector<Frame> up(up_n + OS_PAD * oversample);

    // Prepend context if available
    int pad_out = 0;
    if (_os_context_valid) {
        // Upsample context into start of up[]
        for (int i = 0; i < OS_PAD; ++i) {
            for (int s = 0; s < oversample; ++s) {
                float t = (float)s / oversample;
                Frame a = _os_context[i];
                Frame b = (i+1 < OS_PAD) ? _os_context[i+1] : buf[0];
                up[i * oversample + s] = {lerp(a.l, b.l, t), lerp(a.r, b.r, t)};
            }
        }
        pad_out = OS_PAD * oversample;
    }

    // Upsample main block
    for (int i = 0; i < frames; ++i) {
        Frame next = (i+1 < frames) ? buf[i+1] : buf[frames-1];
        for (int s = 0; s < oversample; ++s) {
            float t = (float)s / oversample;
            up[pad_out + i * oversample + s] = {
                lerp(buf[i].l, next.l, t),
                lerp(buf[i].r, next.r, t)
            };
        }
    }

    // Save context for next block
    _os_context.assign(buf + frames - OS_PAD, buf + frames);
    _os_context_valid = true;

    // Saturate at high sample rate
    int sat_n = pad_out + up_n;
    magnetic.saturate_only(up.data(), sat_n, p);

    // Downsample (averaging decimation)
    for (int i = 0; i < frames; ++i) {
        float l = 0, r = 0;
        int base = pad_out + i * oversample;
        for (int s = 0; s < oversample; ++s) {
            l += up[base + s].l;
            r += up[base + s].r;
        }
        buf[i] = {l / oversample, r / oversample};
    }
}

bool TapeEngine::dsp_process(Frame* out, int frames, int oversample) {
    std::lock_guard<std::mutex> g(lock);

    if (audio_data.empty()) return false;
    // End-of-tape: forward hits end, reverse hits start
    if (!is_reversed && play_head >= total_samples - 1) return false;
    if ( is_reversed && play_head <= 0.0)               return false;

    EngineParams p = params;
    p.is_reversed  = is_reversed;
    // When tape is moving, keep engage target = 1.0 so TransportDynamics
    // internal ramp reaches full speed quickly. tape_speed_mult carries
    // the actual inertia — no double-ramping.
    if (p.tape_speed_mult > 0.001f) p.motor_engage = 1.0f;
    float speed_factor = p.ips_base / 15.0f;

    // Transport
    auto tr = transport.process(frames, current_time, play_head, total_samples, p);
    if (tr.speeds.empty()) return false;

    // Build read indices.
    // When reversed, advance BACKWARD through audio_data so play_head always
    // represents the true forward position in the original file.
    std::vector<double> read_indices(frames);
    double acc = play_head;
    if (p.is_reversed) {
        for (int i = 0; i < frames; ++i) {
            acc -= tr.speeds[i] * p.tape_speed_mult;  // backward at speed_mult
            read_indices[i] = acc;
        }
    } else {
        for (int i = 0; i < frames; ++i) {
            acc += tr.speeds[i] * p.tape_speed_mult;
            read_indices[i] = acc;
        }
    }

    // Catmull-Rom cubic Hermite interpolation — C1 continuous, no zipper noise
    // at any tape speed. 4-19× smoother than linear at slow spindown speeds.
    for (int i = 0; i < frames; ++i) {
        double pos = std::clamp(read_indices[i], 0.0, (double)(total_samples - 1));
        out[i] = cubic_interp(audio_data.data(), total_samples, pos);
    }

    // Oversampled saturation or normal magnetic processing
    if (oversample > 1) {
        _saturate_oversampled(out, frames, oversample, p);
        EngineParams p_os = p;
        p_os.presaturated = true;
        magnetic.process(out, frames,
                         std::span<const Frame>(audio_data),
                         std::span<const double>(read_indices), p_os);
    } else {
        magnetic.process(out, frames,
                         std::span<const Frame>(audio_data),
                         std::span<const double>(read_indices), p);
    }

    electronics.process(out, frames, current_time, speed_factor, tr.sticky_drag, p);

    int n_out = frames;

    // ── Scrape flutter ────────────────────────────────────────────────────────
    float scrape_amt = p.scrape_flutter;
    if (scrape_amt > 0.001f) {
        float ips        = p.ips_base;
        float scrape_hz  = std::clamp(2800.0f * (ips / 15.0f), 600.0f, 14000.0f);
        float phase_inc  = scrape_hz / SR_F;
        float alpha      = std::clamp(0.85f + scrape_amt * 0.1f, 0.85f, 0.97f);
        float depth      = std::clamp(scrape_amt * 0.4f, 0.0f, 0.45f);

        static thread_local std::mt19937 sc_rng{42};
        static thread_local std::normal_distribution<float> sc_norm{0.f, 0.3f};
        static thread_local std::uniform_real_distribution<float> sc_uni{0.f, 1.f};

        float y_prev = _scrape_y_prev;
        float x_prev = _scrape_x_prev;

        for (int i = 0; i < n_out; ++i) {
            float ph = _scrape_phase + i * phase_inc;
            float env = std::abs(std::sin(TWO_PI * ph)
                              + std::sin(TWO_PI * ph * 1.031f + 0.7f) * 0.6f
                              + std::sin(TWO_PI * ph * 0.973f + 1.3f) * 0.4f
                              + sc_norm(sc_rng));
            // Normalise: we just clamp rather than max-normalise per-sample
            env = std::min(env, 2.0f) * 0.5f;

            float x_cur = out[i].l;
            float y_cur = alpha * (y_prev + x_cur - x_prev);
            float mod   = y_cur * env * depth;
            out[i].l += mod;
            out[i].r += mod * 0.85f;
            x_prev = x_cur;
            y_prev = y_cur;
        }
        _scrape_phase = std::fmod(_scrape_phase + n_out * phase_inc, 1.0f);
        _scrape_y_prev = y_prev;
        _scrape_x_prev = x_prev;
    }

    // Dropout mask
    for (int i = 0; i < n_out; ++i) {
        out[i] *= tr.dropout_mask[i];
    }

    // Fighting motors AM
    float conflict = transport.last_conflict;
    if (conflict > 0.05f) {
        float am_freq  = 2.0f + conflict * 3.0f;
        float am_depth = std::clamp(conflict * 0.7f, 0.0f, 0.8f);
        for (int i = 0; i < n_out; ++i) {
            float t   = current_time + (float)i * SR_F_INV;
            float am  = 1.0f - am_depth * (0.5f + 0.5f * std::sin(TWO_PI * am_freq * t));
            out[i] *= am;
        }
    }

    // play_head always tracks the actual tape position in the forward file.
    // Clamp to [0, total_samples-1] in both directions.
    play_head = std::clamp(read_indices.back(), 0.0, (double)(total_samples - 1));
    current_time += (float)frames * SR_F_INV;

    // Output limiter (analog-style soft clip)
    const float KNEE = 0.97f;
    for (int i = 0; i < n_out; ++i) {
        out[i].l = fast_tanh(out[i].l / KNEE) * KNEE;
        out[i].r = fast_tanh(out[i].r / KNEE) * KNEE;
    }

    // Post-preset fade-in
    const int FADE_SAMPS = 512;
    if (_fade_in > 0) {
        int fl = std::min(_fade_in, n_out);
        int pos_start = FADE_SAMPS - _fade_in;
        for (int i = 0; i < fl; ++i) {
            float t = (float)(pos_start + i) / FADE_SAMPS;
            out[i].l = out[i].l * t + _last_out.l * (1.0f - t);
            out[i].r = out[i].r * t + _last_out.r * (1.0f - t);
        }
        _fade_in -= fl;
    }
    _last_out = out[n_out - 1];

    return true;
}

std::unique_ptr<TapeEngine> TapeEngine::make_worker(
    int start_sample, int block_size, int oversample,
    uint64_t shared_seed, int n_warmup_blocks)
{
    auto eng = std::make_unique<TapeEngine>(shared_seed);
    {
        std::lock_guard<std::mutex> g(lock);
        eng->audio_data    = audio_data;
        eng->total_samples = total_samples;
        eng->params        = params;
        eng->is_reversed   = is_reversed;
    }

    // Warm up DSP state before the slice starts.
    // For forward renders: warm up from (start - warmup) advancing forward to start.
    // For reverse renders: warm up from (start + warmup) retreating backward to start.
    int warmup_samples = n_warmup_blocks * block_size;
    eng->transport.motor_engage        = 1.f;
    eng->transport.current_motor_speed = 1.f;

    std::vector<Frame> discard(block_size);
    if (!eng->is_reversed) {
        int warmup_start = std::max(0, start_sample - warmup_samples);
        eng->play_head   = (double)warmup_start;
        eng->current_time = (float)warmup_start / SR_F;
        while (eng->play_head < (double)start_sample) {
            if (!eng->dsp_process(discard.data(), block_size, oversample)) break;
        }
    } else {
        int warmup_start = std::min(eng->total_samples - 1, start_sample + warmup_samples);
        eng->play_head   = (double)warmup_start;
        eng->current_time = (float)warmup_start / SR_F;
        while (eng->play_head > (double)start_sample) {
            if (!eng->dsp_process(discard.data(), block_size, oversample)) break;
        }
    }

    eng->play_head    = (double)start_sample;
    eng->current_time = (float)start_sample / SR_F;
    return eng;
}
